### PR TITLE
Inventory set operations

### PR DESCRIPTION
**Pull request by Faithcaio**
*SpongeAPI* | [SpongeCommon](https://github.com/SpongePowered/SpongeCommon/pull/1451)

Adds `intersect` `union` and `containsInventory` to the Inventory API

Issue: https://github.com/SpongePowered/SpongeAPI/issues/1615

@Mumfrey

### CODE REVIEW

**Comment by ryantheleach** [comment]
@Mumfrey What about hotbar.except(first) ? where hotbar is the hotbar inventory, and first is the first hotbar slot?

(vs using the term complement, which is well known to Set theorists, but except is more known in SQL among programmers)

File: src/main/java/org/spongepowered/api/item/inventory/Inventory.java [hunk 1]
```diff
@@ -505,6 +505,33 @@
...
+    Inventory union(Inventory inventory);
+
+    /**
+     * Returns true if the given inventory is a descendant of this one.
+     * This includes deeply nested ones but not usually not query-results.
+     *
+     * @param inventory the other inventory
+     * @return whether the given inventory is contained in this one.
+     */
+    boolean containsInventory(Inventory inventory);
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/item/inventory/Inventory.java:544]: I would rephrase or extend the javadoc by something like:
* `an inventory is contained in another if all slots are contained in the parent`
* `an inventory is a decendant of an inventory if all slots of the decendant are contained in the parent`

This method can also be default implemented using:
`this.union(inventory).capacity() == this.capacity()`

Comment (Faithcaio) [reply src/main/java/org/spongepowered/api/item/inventory/Inventory.java:544]: ContainsInventory will not operate on each slot. 
So you usually using queryresults as parameter will return false.
Except if you check e.g. for a row to be contained in a gridinventory
or a single slot in an ordered inventory

File: src/main/java/org/spongepowered/api/item/inventory/Inventory.java [hunk 2]
```diff
@@ -505,6 +505,33 @@
     InventoryArchetype getArchetype();
 
     /**
+     * Intersects the slots of both inventories
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/item/inventory/Inventory.java]: This needs more documentation as it isn't clear what is going on whatsoever here.

File: src/main/java/org/spongepowered/api/item/inventory/Inventory.java [hunk 3]
```diff
@@ -505,6 +505,35 @@
...
+     * from both inventories.
+     *
+     * @param inventory the other inventory
+     * @return an inventory wrapping all slots of both inventories.
+     */
+    Inventory union(Inventory inventory);
+
+    /**
+     * Returns true if the given inventory is a descendant of this one.
+     * This includes deeply nested ones but not usually not query-results.
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/item/inventory/Inventory.java]: This needs further explanation...

It doesn't affect queries or?

Comment (Faithcaio) [reply src/main/java/org/spongepowered/api/item/inventory/Inventory.java]: Maybe @Mumfrey can find a good way to describe it?

Comment (ryantheleach) [reply src/main/java/org/spongepowered/api/item/inventory/Inventory.java]: Query Results are essentially a new Inventory View created to solve the query.

The Query Result returned, may, or may not be, a sub inventory of an existing inventory depending on the query, or it may be an amalgamation from various sources.

If you have a reference to a sub-inventory, this will work in contains.

However, after sleeping on this after we talked @Faithcaio I'm not sure this is entirely expected from user-developers.

Maybe @ST-DDT Is correct, in that we should essentially run contains for all of the smallest inventory views contained within the query result, as I think there are situations where you could end up with a custom view of a compound inventory made of multiple slots, that all belong to a single parent.

Unless you want to change the signature of query results to be a collection like class of inventories anyway, to make this explicit.

Comment (Faithcaio) [reply src/main/java/org/spongepowered/api/item/inventory/Inventory.java]: @ryantheleach if you wanted to operate on slots you would just use the intersect method

**Comment by kashike** [comment]
https://github.com/SpongePowered/SpongeAPI/commit/4e8d9e92cb4ff4a4c3a877dfcd454a64db8d3f92
